pkg/testutils/perfring/perfring.go

```diff
@@ -55,37 +72,69 @@ func ProcessEvents(t *testing.T, ctx context.Context, eventFn EventFn, wgStarted
 	if err != nil {
 		t.Fatalf("creating perf array reader failed: %v", err)
 	}
+
 	wgStarted.Done()
 
-	complChecker := testsensor.NewCompletionChecker()
-	for {
-		if ctx.Err() != nil {
-			break
-		}
+	errChan := make(chan error)
+	defer close(errChan)
 
-		record, err := perfReader.Read()
-		if err != nil {
-			t.Fatalf("error reading perfring buffer: %v", err)
-		}
+	complChan := make(chan bool)
+	defer close(complChan)
 
-		events, err := observer.HandlePerfData(record.RawSample)
-		if err != nil {
-			t.Fatalf("error handling perfring data: %v", err)
-		}
-		for _, ev := range events {
-			switch xev := ev.(type) {
-			case *testapi.MsgTestEventUnix:
-				cpu := xev.Arg0
-				complChecker.Update(cpu)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Wait()
+
+	go func() {
+		defer wg.Done()
+
+		complChecker := testsensor.NewCompletionChecker()
+
+		for {
+			if ctx.Err() != nil {
+				break
+			}
+
+			record, err := perfReader.Read()
+			if err != nil {
+				if ctx.Err() == nil {
+					errChan <- fmt.Errorf("error reading perfring data: %v", err)
+				}
+				break
+			}
+
+			events, err := observer.HandlePerfData(record.RawSample)
+			if err != nil {
+				errChan <- fmt.Errorf("error handling perfring data: %v", err)
+				break
 			}
-			if err := eventFn(ev); err != nil {
-				t.Fatalf("event handling function returned: %s", err)
+			err = loopEvents(events, eventFn, complChecker)
+			if err != nil {
+				errChan <- fmt.Errorf("error loop event function returned: %s", err)
+				break
+			}
+
+			if complChecker.Done() {
+				complChan <- true
+				break
 			}
 		}
-		if complChecker.Done() {
-			break
+	}()
+
+	for {
+		select {
+		case err := <-errChan:
+			t.Fatal(err)
```

Comment:
Any reason why we are calling Fatal here?

Reply:
Original code was calling fatal , didn't change semantics, just added the missing sync logic